bolts/data/datasets/utils.py

```diff
@@ -78,3 +97,180 @@ def apply_image_transform(
                 image = Image.fromarray(image)
             image_ = transform(image)
     return image_
+
+
+def img_to_tensor(img: Image.Image | np.ndarray) -> Tensor:
+    if isinstance(img, Image.Image):
+        return F.pil_to_tensor(img)
+    return torch.from_numpy(
+        np.moveaxis(img / (255.0 if img.dtype == np.uint8 else 1), -1, 0).astype(np.float32)
+    )
+
+
+@overload
+def extract_base_dataset(
+    dataset: Dataset, return_subset_indices: Literal[True] = ...
+) -> tuple[Dataset, Tensor | slice]:
+    ...
+
+
+@overload
+def extract_base_dataset(dataset: Dataset, return_subset_indices: Literal[False] = ...) -> Dataset:
+    ...
+
+
+def extract_base_dataset(
+    dataset: Dataset, return_subset_indices: bool = True
+) -> Dataset | tuple[Dataset, Tensor | slice]:
+    def _closure(
+        dataset: Dataset, rel_indices_ls: list[list[int]] | None = None
+    ) -> Dataset | tuple[Dataset, Tensor | slice]:
+        if rel_indices_ls is None:
+            rel_indices_ls = []
+        if hasattr(dataset, "dataset"):
+            if isinstance(dataset, Subset):
+                rel_indices_ls.append(list(dataset.indices))
+            return _closure(dataset.dataset, rel_indices_ls)  # type: ignore
+        if return_subset_indices:
+            if rel_indices_ls:
+                abs_indices = torch.as_tensor(rel_indices_ls.pop(), dtype=torch.long)
+                for indices in rel_indices_ls[::-1]:
+                    abs_indices = abs_indices[indices]
+            else:
+                abs_indices = slice(None)
+            return dataset, abs_indices
+        return dataset
+
+    return _closure(dataset)
+
+
+@lru_cache(typed=True)
+def extract_labels_from_dataset(dataset: Dataset) -> tuple[Tensor | None, Tensor | None]:
+    """Attempt to extract s/y labels from a dataset."""
+
+    def _closure(dataset: Dataset) -> tuple[Tensor | None, Tensor | None]:
+        dataset, indices = extract_base_dataset(dataset, return_subset_indices=True)
+        _s = None
+        _y = None
+        if getattr(dataset, "s", None) is not None:
+            _s = dataset.s[indices]  # type: ignore
+        if getattr(dataset, "y", None) is not None:
+            _s = dataset.s[indices]  # type: ignore
+
+        _s = torch.from_numpy(_s) if isinstance(_s, np.ndarray) else _s
+        _y = torch.from_numpy(_y) if isinstance(_y, np.ndarray) else _y
+
+        return _s, _y
+
+    if isinstance(dataset, (ConcatDataset)):
+        s_all_ls, y_all_ls = [], []
```

Comment:
For typing, I think these will have to go over two line
```suggestion
        s_all_ls: list[Tensor] = []
        y_all_ls: list[Tensor] = []
```